File: Matlab/DMSToDMS.m
% 无损精度提取度分秒函数
function dms = DMSToDMS(dmsAngle)
    dmsAngle = dmsAngle * 10000; % 为了保证精度,先乘以10000
    % fix表示截断取整，即直接抹去小数部分
    sAngle = fix(dmsAngle); % 取整数部分进行提取
    d = fix(sAngle/10000); % 获取度
    m = fix((sAngle - d * 10000)/100); % 获取分
    s = dmsAngle - d * 10000 - m * 100; % 获取秒
    dms = [d, m, s]; % 返回向量
end

File: Python/DMSToDMS.py
# -*- coding: utf-8 -*-
# @Time    :2023/12/23 23:28
# @Author  :小 y 同 学
# @公众号   :小y只会写bug
# @CSDN    :https://blog.csdn.net/weixin_64989228?type=blog
def DMSToDMS(dmsAngle: float) -> list:
	"""
	无精度损失的提取方法
	:param dmsAngle: 浮点数度分秒如1.314520表示1度13分14.520秒
	:return: list [度,分,秒]
	"""
	dmsAngle *= 10000  # 为了保证精度,先乘以10000
	sAngle = int(dmsAngle)  # 取整数部分进行提取
	d = int(sAngle / 10000)  # 获取度int
	m = int((sAngle - d * 10000) / 100)  # 获取分int
	s = dmsAngle - d * 10000 - m * 100  # 获取秒 float
	return [d, m, s]


if __name__ == '__main__':
	a = 1.40000
	print(DMSToDMS(float(a)))
	b = 13.14520
	print(DMSToDMS(float(b)))
